tmt/steps/__init__.py

```diff
@@ -392,6 +404,135 @@ def go(self):
+
+        # Run the interactive command
+        self.info('reboot', 'Rebooting machine', color='yellow')
+        for guest in self.parent.plan.provision.guests():
+            # Attempt to push the workdir to the guest
+            try:
+                guest.push()
+                cwd = self.parent.plan.worktree
+            except tmt.utils.GeneralError:
+                self.warn("Failed to push workdir to the guest.")
+                cwd = None
+            guest.reboot(self._hard)
+        self.info('reboot', 'Reboot finished', color='yellow')
+
+    @classmethod
+    def phases(cls, step):
+        """ Return list of phases enabled for given step """
+        # Build the phase list unless done before
+        if cls._phases is None:
+            cls._phases = cls._parse_phases(step)
+        # Return enabled phases, empty list if step not found
+        try:
+            return cls._phases[step.name]
+        except KeyError:
+            return []
+
+    @classmethod
+    def _parse_phases(cls, step):
+        """ Parse options and store phase order """
+        phases = dict()
+        options = cls._opt('step')
+
+        # Use the end of the last enabled step if no --step given
+        if not options:
+            last_enabled_step = list(step.plan.steps())[-1]
+            phases[last_enabled_step.name] = [PHASE_END]
+
+        # Process provided options
+        for option in options:
+            # Parse the step:phase format
+            matched = re.match(r'(\w+)(:(\w+))?', option)
+            if matched:
+                step_name, _, phase = matched.groups()
+            if not matched or step_name not in STEPS:
+                raise tmt.utils.GeneralError(f"Invalid step '{option}'.")
+            # Check phase format, convert into int, use end by default
+            try:
+                phase = int(phase)
+            except TypeError:
+                phase = PHASE_END
+            except ValueError:
+                # Convert 'start' and 'end' aliases
+                try:
+                    phase = dict(start=PHASE_START, end=PHASE_END)[phase]
+                except KeyError:
+                    raise tmt.utils.GeneralError(f"Invalid phase '{phase}'.")
+            # Store the phase for given step
+            try:
+                phases[step_name].append(phase)
+            except KeyError:
+                phases[step_name] = [phase]
+        return phases
+
+    @classmethod
+    def phases(cls, step):
+        """ Return list of phases enabled for given step """
+        # Build the phase list unless done before
+        if cls._phases is None:
+            cls._phases = cls._parse_phases(step)
+        # Return enabled phases, empty list if step not found
+        try:
+            return cls._phases[step.name]
+        except KeyError:
+            return []
+
+    @classmethod
+    def plugins(cls, step):
+        """ Return list of login instances for given step """
+        if not Login._enabled:
```

Comment:
@RHEmployee the problem why the reboot subcommand is not invoked is here. Copy-paste error - you are checking on the `Login` class rather than the `Reboot` class and `Login` obviously is not enabled but `Reboot` is.

The question is - do we really need phases for reboot? I am not completely sure but I guess it might be useful to specify when exactly the reboot should be done, it gives more control. If we agree that we want them, the common logic for their handling should probably be in a class that both `Login` and `Reboot` will inherit from

Reply:
Hi! Thanks! I wasn't able to see it. Too much iteration of looking at code again and again.
Reproducer is same as provided command (without fix). However, now, when I can move forward, I have different issue with phases and options so I need to take a look what's going on.

For example:
  File "/home/rhack/devel/tmt/tmt/steps/discover/__init__.py", line 77, in wake
    super(Discover, self).wake()
  File "/home/rhack/devel/tmt/tmt/steps/__init__.py", line 148, in wake
    for plugin in Reboot.plugins(step=self):
  File "/home/rhack/devel/tmt/tmt/steps/__init__.py", line 533, in plugins
    return [Reboot(step, phase) for phase in cls.phases(step)]
  File "/home/rhack/devel/tmt/tmt/steps/__init__.py", line 521, in phases
    cls._phases = cls._parse_phases(step)
  File "/home/rhack/devel/tmt/tmt/steps/__init__.py", line 491, in _parse_phases
    for option in options:
TypeError: 'NoneType' object is not iterable

But I don't know anything about phases code, and how it work, for now.